### ISSUE TITLE
rocksdb built with gcc-7.3.0 and mingw-w64-5.0.3 hits assertion

### ISSUE THREAD

**Issue by angerman** [facebook/rocksdb#3984]
So I've built rocksdb-shared on macOS using gcc/mingw-w64 from brew.

This builds the `librocksdb-shared.dll` just fine, but when using it, instead of
a prebuilt `rocksdb.dll` with msvc, it hits this assert here:

https://github.com/facebook/rocksdb/blob/f9e0a0651244f9cea60e9fffadb73c27ee850ff5/db/column_family.cc#L477

I speculate that this is due to using libwinpthreads and the thread local storage destructors in https://github.com/facebook/rocksdb/blob/a6d8e30c05ac1a2cd4bfa9aa7d95a498b02b354c/util/thread_local.cc with lots of `OS_WIN` ifdefs. I could be wrong though.

**Comment by sagar0** [comment]
I didn't fully understand your setup here. How are you building a dll on os x? Are you saying that you are cross-compiling on os x using gcc/mingw to get a dll?

**Comment by angerman** [comment]
Yes. I’m using gcc/mingw with winpthreads to compile rocksdb on macOS to produce a Windows dll.

Quite similar to how the mingw Travis CI Job runs. It compiles fine, just seems to have different runtime behaviour to the dll compiled with MSVC on windows.

**Comment by yiwu-arbug** [comment]
@angerman Can you paste a sample code of how you using rocksdb and trigger the assertion?
